### DISCUSSION TITLE
Why do unsupervised transformers only update on predict and not also on learn?

### DISCUSSION THREAD

**Discussion by TheRisenPhoenix** [online-ml/river#542] [Q&A] [answered]
I came across an issue when I wanted to do a pretraining of the model, it would fail to learn anything. After quite some time I figured out that the Standardscaler I use in my pipeline (Standardscaler -> PAClassifier) would always output 0/0/0/0... for all features. Thanks to the documentation and a couple of issues/discussions here, I realized that this is the desired behaviour, unsupervised transformers should only update when calling _predict_, not when calling _learn_:

>  One special thing to take notice to is the way transformers are handled. In a typical scenario,
    it is usual to predict something for a sample and wait for the ground truth to arrive. In such
    a case, the features are seen before the ground truth arrives. Therefore, the unsupervised
    parts of the pipeline are updated when `predict_one` and `predict_proba_one` are called.
    Usually the unsupervised parts of the pipeline are all the steps that precede the final step,
    which is a supervised model. However, some transformers are supervised and are therefore
    obtained during calls to `learn_one`.

I see that you would like to update transformers when _predicting_, but I don't understand why they would not also be updated when calling _learn_. I'm not judging, just asking, so it would be great if someone could explain this to me.
Also, is there any official way to accomplish pretraining? I could call _predict_() before each _learn_(), but that doesn't feel that right.

**Comment by MaxHalford** [comment, marked as answer]
It's a good question and comes up a lot.

The reason why we update the transformers in `predict_one` is because we have all the information we need at that point, and it performs better to update the transformers as soon as possible. This is especially true for transformers. If we update the transformers in `learn_one`, we could be updating them twice, which is not desirable.

Indeed the current way to do pretraining is to `predict_one` before `learn_one`. I understand it's not ideal. What we could do is add a `learn_unsupervised` boolean parameter to the `learn_one` method.

**Reply by TheRisenPhoenix** [reply]
Thank you for the explanation.

> If we update the transformers in learn_one, we could be updating them twice, which is not desirable.

I see. Updating the transformers multiple times might result in unnecessary computation as well as unbalanced adjustments (using trainingdata twice results in a false view on the data, when predictiondata is only used once to update). Did I catch that correctly?

> What we could do is add a learn_unsupervised boolean parameter to the learn_one method.

Sounds good. I think that would help a lot.
For the moment, I will go with the predict->learn solution. 

Thank you!

**Reply by MaxHalford** [reply]
This is adressed in #545.
